compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs

```diff
@@ -310,25 +311,40 @@ where
         // normalizing the self type as well, since type variables are not uniquified.
         let goal = self.resolve_vars_if_possible(goal);
 
-        let mut candidates = vec![];
-
         if let TypingMode::Coherence = self.typing_mode() {
             if let Ok(candidate) = self.consider_coherence_unknowable_candidate(goal) {
                 return vec![candidate];
             }
         }
 
-        self.assemble_impl_candidates(goal, &mut candidates);
-
-        self.assemble_builtin_impl_candidates(goal, &mut candidates);
-
-        self.assemble_alias_bound_candidates(goal, &mut candidates);
-
-        self.assemble_object_bound_candidates(goal, &mut candidates);
+        assemble(self, goal)
+    }
 
-        self.assemble_param_env_candidates(goal, &mut candidates);
+    pub(super) fn assemble_and_evaluate_candidates_from_env_and_bounds<G: GoalKind<D>>(
+        &mut self,
+        goal: Goal<I, G>,
+    ) -> Vec<Candidate<I>> {
+        self.assemble_and_evaluate_candidates_with_normalized_goal(goal, |ecx, goal| {
+            let mut candidates = vec![];
+            ecx.assemble_alias_bound_candidates(goal, &mut candidates);
```

Comment:
Bit unsatisfying that we have to keep this in sync w/ below. Could introduce another helper function, I guess.